client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -186,8 +191,143 @@ private Checkpoint completeCheckpoint(String checkpointName) {
+        val config = state.getConfig();
+        val generation = state.getGeneration();
+        long segment = synchronizer.getSegmentId();
+        if (!ReaderGroupConfig.RetentionConfig.NONE.equals(config.getRetentionConfig())) {
+            Set<Stream> streams = config.getStartingStreamCuts().keySet();
+            streams.forEach(s -> getThrowingException(controller.addSubscriber(scope, s.getStreamName(), groupName + segment, generation)));
+        }
+        synchronizer.updateState((s, updates) -> {
+            if (s.equals(state)) {
+                updates.add(new ChangeConfigState(ReaderGroupState.ConfigState.READY, s.getGeneration()));
+            }
+        });
+    }
+
+    public void doReinit(ReaderGroupState state) {
+        val oldConfig = state.getConfig();
+        val newConfig = state.getNewConfig();
+        val generation = state.getGeneration();
+        manageSubscriptions(oldConfig, newConfig, generation);
+
+        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
+        synchronizer.updateState((s, updates) -> {
+            if (s.equals(state)) {
+                updates.add(new ReaderGroupStateResetComplete(newConfig, segments, getEndSegmentsForStreams(newConfig)));
+            }
+        });
+    }
+
+    public void doDelete(ReaderGroupState state) {
+        val config = state.getConfig();
+        val generation = state.getGeneration();
+        long segment = synchronizer.getSegmentId();
+        if (!ReaderGroupConfig.RetentionConfig.NONE.equals(config.getRetentionConfig())) {
+            Set<Stream> streams = config.getStartingStreamCuts().keySet();
+            streams.forEach(s -> getThrowingException(controller.deleteSubscriber(scope, s.getStreamName(), groupName + segment, generation)));
+        }
+    }
+
+    private void manageSubscriptions(ReaderGroupConfig oldConfig, ReaderGroupConfig newConfig, long generation) {
+        long segment = synchronizer.getSegmentId();
+        Set<Stream> oldStreams = oldConfig.getRetentionConfig() != ReaderGroupConfig.RetentionConfig.NONE ? oldConfig.getStartingStreamCuts().keySet() : Collections.emptySet();
+        Set<Stream> newStreams = newConfig.getRetentionConfig() != ReaderGroupConfig.RetentionConfig.NONE ? newConfig.getStartingStreamCuts().keySet() : Collections.emptySet();
+        Set<String> streamsToSub = filterOut(newStreams, oldStreams).stream().map(Stream::getStreamName).collect(Collectors.toSet());
+        Set<String> streamsToUnsub = filterOut(oldStreams, newStreams).stream().map(Stream::getStreamName).collect(Collectors.toSet());
+        streamsToSub.forEach(s -> getThrowingException(controller.addSubscriber(scope, s, groupName + segment, generation)));
+        streamsToUnsub.forEach(s -> getThrowingException(controller.deleteSubscriber(scope, s, groupName + segment, generation)));
+    }
+
+    private boolean updateConfigState(ReaderGroupState.ConfigState configState, ReaderGroupState state) {
```

Comment:
this method looks identical to the code above in resetrg

Reply:
I've added another method which takes the `ReaderGroupStateUpdate` object to be applied.